### PR TITLE
Implement `pos_calc` module

### PR DESCRIPTION
**Pull request by ericonr**

### CODE REVIEW

**Comment by ericonr** [comment]
It is still necessary to improve the `wait_fifo` interface, especially for non-blocking users.

File: modules/pos_calc.cc [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        /*uint32_t t = *rate.tag;
+        add_channel("SYNC_EN", i, get_bit(t, POS_CALC_TBT_TAG_EN));
+        add_channel("SYNC_DLY", i, extract_value<uint16_t>(t, POS_CALC_TBT_TAG_DLY_MASK));
+        add_channel("DESYNC_CNT", i, extract_value<uint16_t>(t, POS_CALC_TBT_TAG_DESYNC_CNT_MASK));
+
+        add_channel("DATA_MASK_EN", i, get_bit(*rate.data_mask_ctl, POS_CALC_TBT_DATA_MASK_CTL_EN));
+
+        t = *rate.data_mask_samples;
+        add_channel("DATA_MASK_SAMPLES_BEG", i, extract_value<uint16_t>(t, POS_CALC_TBT_DATA_MASK_SAMPLES_BEG_MASK));
+        add_channel("DATA_MASK_SAMPLES_END", i, extract_value<uint16_t>(t, POS_CALC_TBT_DATA_MASK_SAMPLES_END_MASK));*/
```
Comment (ericonr) [inline modules/pos_calc.cc:253]: Forgot to remove this.

**Comment by ericonr** [comment]
`write_params` has to be fully replaced to account for the AMPFIFO behavior, since all bar4 writes are accompanied by a read. Thankfully, the register is ro, so we can simply skip it.

**Comment by ericonr** [comment]
We will replace the `Controller` implementation with #41 before merging.

File: modules/pos_calc.cc [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    add_general("AMPFIFO_MONIT_EMPTY", get_bit(t, POS_CALC_AMPFIFO_MONIT_AMPFIFO_MONIT_CSR_EMPTY));
+}
+
+void Core::decode_fifo_amps()
+{
+    add_channel("AMPFIFO_MONIT_AMP", 0, regs.ampfifo_monit.ampfifo_monit_r0);
+    add_channel("AMPFIFO_MONIT_AMP", 1, regs.ampfifo_monit.ampfifo_monit_r1);
+    add_channel("AMPFIFO_MONIT_AMP", 2, regs.ampfifo_monit.ampfifo_monit_r2);
+    add_channel("AMPFIFO_MONIT_AMP", 3, regs.ampfifo_monit.ampfifo_monit_r3);
+}
```
Comment (ericonr) [inline modules/pos_calc.cc:195]: I'm investigating whether I can safely change these values from one thread while another thread does other stuff (assuming I initialize the class by adding them to the map): https://github.com/Tessil/ordered-map/issues/46

**Review by henriquesimoes — COMMENTED** [review]
I'd like to better understand the interface the pos_calc hardware module is providing us to be able to better review `pos_calc.cc`.

At a first look, I didn't find any issues.

File: include/modules/pos_calc.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2023 CNPEM (cnpem.br)
```
[the remark this answers is not in the document]
Comment (ericonr) [reply include/modules/pos_calc.h:2]: Honestly considering removing them all instead. Current wisdom about copyright says it's not necessary.

**Review by henriquesimoes — COMMENTED** [review]
I still cannot feel that my review properly covers `pos_calc.cc`. We may discuss it a bit (in person) if you will.

File: util/tests/decoders-test.cc [hunk 4]
```diff
@@ -182,6 +182,14 @@ TEST_CASE("Channel data single channel", "[decoders-test]")
     CHECK_THROWS_AS(dec.get_channel_data<double>("C_DOUBLE_CH1", 0), std::out_of_range);
 }
 
+TEST_CASE("Generic data API", "[decoders-test]")
+{
+    TestRegisterDecoder dec{};
+    dec.decode();
+
+    CHECK(std::get<int32_t>(dec.get_generic_data("INT")) == 1);
```
[the remark this answers is not in the document]
Comment (ericonr) [reply util/tests/decoders-test.cc:190]: Definitely!

**Comment by ericonr** [comment]
@henriquesimoes there is an [HTML doc](https://github.com/lnls-dig/bpm-gw/blob/b26def03d8e2eaa145c82b5960cba3cb8f469b82/hdl/modules/wb_position_calc/cheby/doc/wb_pos_calc_regs_wb.html) for the register map, but I don't think it actually counts as explaining anything. We can get together sometime next week to go over it properly :)

File: modules/pos_calc.cc [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    auto rf_adc_gains_fixed_point_pos = rf_extract_value(regs.adc_gains_fixed_point_pos, POS_CALC_ADC_GAINS_FIXED_POINT_POS_DATA_MASK);
+    add_general("ADC_GAINS_FIXED_POINT_POS", rf_adc_gains_fixed_point_pos);
+    auto adc_gains_fixed_point_pos = std::get<int32_t>(rf_adc_gains_fixed_point_pos.value);
+
+    const GainArrays gains_arrays{regs};
+    auto get_gains = [this, adc_gains_fixed_point_pos](const char *name, const auto &gains) {
+        for (unsigned i = 0; i < gains.size(); i++)
+            add_channel(
+                name, i,
+                rf_fixed2float(rf_whole_register(*gains[i]), adc_gains_fixed_point_pos));
```
Comment (henriquesimoes) [inline modules/pos_calc.cc:156]: I've been thinking a bit about this. How does changing the `ADC_GAINS_FIXED_POINT_POS` property will update this cached fixed point position metadata?

We should allow this to be changed at runtime, right?

Comment (ericonr) [reply modules/pos_calc.cc:156]: https://github.com/lnls-dig/bpm-gw/pull/91

**Pull request merged**